lib/IRGen/IRGenModule.cpp

```diff
@@ -1185,63 +1175,179 @@ static bool isFirstObjectFileInModule(IRGenModule &IGM) {
+        IGM.Module.appendModuleInlineAsm(".section .swift1_autolink_entries,"
+                                         "\"0x80000000\"");
+      auto var =
+          new llvm::GlobalVariable(*IGM.getModule(), EntriesConstant->getType(),
+                                   true, llvm::GlobalValue::PrivateLinkage,
+                                   EntriesConstant, "_swift1_autolink_entries");
+      var->setSection(".swift1_autolink_entries");
+      var->setAlignment(llvm::MaybeAlign(IGM.getPointerAlignment().getValue()));
+
+      disableAddressSanitizer(IGM, var);
+      IGM.addUsedGlobal(var);
+      break;
+    }
+    }
+  }
+
+  static AutolinkKind create(const SwiftTargetInfo &TargetInfo,
+                             llvm::Triple Triple,
+                             IRGenLLVMLTOKind LLVMLTOKind) {
+    // When performing LTO, we always use lld that supports auto linking
+    // mechanism with ELF. So embed dependent libraries names in
+    // "llvm.dependent-libraries" instead of "llvm.linker.options".
+    if (TargetInfo.OutputObjectFormat == llvm::Triple::ELF &&
+        LLVMLTOKind != IRGenLLVMLTOKind::None) {
+      return AutolinkKind::LLVMDependentLibraries;
+    } else if (doesTargetAutolinkUsingAutolinkExtract(TargetInfo, Triple)) {
+      return AutolinkKind::SwiftAutoLinkExtract;
+    } else {
+      return AutolinkKind::LLVMLinkerOptions;
+    }
+  }
+};
+} // namespace
+
 void IRGenModule::emitAutolinkInfo() {
+
+  auto Autolink =
+      AutolinkKind::create(TargetInfo, Triple, IRGen.Opts.LLVMLTOKind);
+
+  StringRef AutolinkSectionName = Autolink.getSectionNameMetadata();
+
+  auto *Metadata = Module.getOrInsertNamedMetadata(AutolinkSectionName);
+  llvm::SmallVector<llvm::MDNode *, 4> Entries;
+
   // Collect the linker options already in the module (from ClangCodeGen).
-  // FIXME: This constant should be vended by LLVM somewhere.
-  auto *Metadata = Module.getOrInsertNamedMetadata("llvm.linker.options");
   for (llvm::MDNode *LinkOption : Metadata->operands())
-    AutolinkEntries.push_back(LinkOption);
+    Entries.push_back(LinkOption);
 
-  // Remove duplicates.
-  llvm::SmallPtrSet<llvm::MDNode *, 4> knownAutolinkEntries;
-  AutolinkEntries.erase(std::remove_if(AutolinkEntries.begin(),
-                                       AutolinkEntries.end(),
-                                       [&](llvm::MDNode *entry) -> bool {
-                                         return !knownAutolinkEntries.insert(
-                                                   entry).second;
-                                       }),
-                        AutolinkEntries.end());
-
-  const bool AutolinkExtractRequired =
-      (TargetInfo.OutputObjectFormat == llvm::Triple::ELF && !Triple.isPS4()) ||
-      TargetInfo.OutputObjectFormat == llvm::Triple::Wasm ||
-      Triple.isOSCygMing();
-
-  if (!AutolinkExtractRequired) {
+  llvm::LLVMContext &ctx = Module.getContext();
+
+  // Construct autolinking entries
+  switch (Autolink) {
```

Comment:
Can you sink this into the AutolinkKind type. I shouldn't see any switches inline here.

Reply:
👍